addons/mail/static/src/messaging/component/notification_list/notification_list_notification_group_tests.js

```diff
@@ -0,0 +1,629 @@
+    // they are linked to different channels, even though the model is the same.
+    assert.expect(7);
+
+    this.data.initMessaging.mail_failures = {
+        11: {
+            last_message_date: moment.utc().format("YYYY-MM-DD HH:mm:ss"),
+            message_id: 11,
+            message_type: 'email',
+            model: 'mail.channel', // key element of this test: `mail.channel`
+            model_name: "Channel",
+            notifications: {
+                21: {
+                    failure_type: 'SMTP',
+                    notification_id: 21,
+                    notification_status: 'exception',
+                    notification_type: 'email',
+                    partner_id: 41,
+                    partner_name: "Someone",
+                },
+            },
+            res_id: 31, // key element of this test: a different res_id
+        },
+        12: {
+            last_message_date: moment.utc().format("YYYY-MM-DD HH:mm:ss"),
+            message_id: 12,
+            message_type: 'email',
+            model: 'mail.channel', // key element of this test: `mail.channel`
+            model_name: "Channel",
+            notifications: {
+                22: {
+                    failure_type: 'SMTP',
+                    notification_id: 22,
+                    notification_status: 'bounce',
+                    notification_type: 'email',
+                    partner_id: 42,
+                    partner_name: "Someone else",
+                },
+            },
+            res_id: 32, // key element of this test: a different res_id
+        },
+    };
+
+    await this.start({
+        async mockRPC(route, args) {
+            if (args.method === 'channel_fold') {
+                assert.step(`rpc:${args.method}/${args.kwargs.state}`);
+                return [];
+            }
+            return this._super(...arguments);
+        },
+    });
+    await this.createNotificationListComponent();
+
+    assert.containsN(
+        document.body,
+        '.o_NotificationGroup',
+        2,
+        "should have 2 notifications group"
+    );
+    const groups = document.querySelectorAll('.o_NotificationGroup');
+    assert.containsOnce(
+        groups[0],
+        '.o_NotificationGroup_counter',
+        "should have 1 group counter in first group"
+    );
+    assert.strictEqual(
+        groups[0].querySelector('.o_NotificationGroup_counter').innerText,
+        "(1)",
+        "should have 1 notification in first group"
+    );
+    assert.containsOnce(
+        groups[1],
+        '.o_NotificationGroup_counter',
+        "should have 1 group counter in second group"
+    );
+    assert.strictEqual(
+        groups[1].querySelector('.o_NotificationGroup_counter').innerText,
+        "(1)",
+        "should have 1 notification in second group"
+    );
+
+    document.querySelector('.o_NotificationGroup').click();
+    assert.verifySteps(['rpc:channel_fold/open']);
```

Comment:
not fond of RPC assertion, prefer assertion that chat window is visible

Reply:
This is a component test for `NotificationList`, it shouldn't care about the chat window. I don't want to pass `hasChatWindow` to `start` if not necessary.

Reply:
IMO check presence of chat window is a much better assertion than relying on server-side synchronization feature of chat windows...

Reply:
I think both are kinda bad, as the only thing the component being tested is aware about this click is "open thread". This unit test shouldn't even make a distinction between open in discuss or open in chat window...

But I don't see how it would be possible to test this without patching the entity to assert.step inside thread open.

Reply:
Maybe this highlight a bigger problem with that thread open (that we already know, for other reasons), and if we want to do such thing about it, maybe it should be a function in the env so we can replace it, or maybe it should be a trigger so we can catch it?

Reply:
Agree both solutions are not perfect, but just wanted to point out that it's not a reason to not choose the much preferred one.

This is a component test and the aim of click is to open the conversation, be it on chat window, discuss, form view, etc... it doesn't care much... Could have a standardize `data-` that identifies a conversation in use by any component. Ideally not relying on technical stuff (so not local-id) but something understandable by humans and could be on UI but isn't just because it shouldn't be in the highlighted. If we take the analogy of a component being a painting, then this `data-` is like a signature.

Reply:
Well the preferred one to me is the one with the RPC, or even doing a patch of open (to highlight a design issue). I prefer those than instantiating another component that is not required, if we want to clean test utils start, this is one step in the right direction, not introducing more uses of `hasChatWindow` and other similar config.

In component tests, only `ChatWindowManager` tests should instantiate itself.

Reply:
I disagree with preferring assertion on chat window server-sync over assertion on chat window presence...

I think your reasoning comes from not having to instantiate chat window manager in test if we don't have, which is great in theory but in practice if you rely on RPC `channel_fold`, it's worse:

Technically, `start()` with `hasChatWindow` should also disable all logic regarding chat windows, so `channel_fold` shouldn't have been called neither. The reason it does is because we were lazy with this param and just limit with mount of chat window manager (and the fact we plan refactoring `start()` made this improvement kinda pointless in short-term).

I really think the lesser evil of most straightforward approach to make a not-so-bad test is to mount chat window manager and assert presence of chat windows. Not mounting it and asserting on RPC makes this test pass by circumstance, and it's just crying for failure as soon as it can in the future.

That's why I much prefer to consider this as an integration test between notification_group_list and chat_window_manager. In that case, it's better to make the most obvious assertion on a chat window being open, and it's definitely it's presence in DOM over the `channel_fold` RPC.